### ISSUE TITLE
NaN weights after update with half precision

### ISSUE THREAD

**Issue by jafioti** [coreylowman/dfdx#748]
When I use half precision, I run into an issue where the weights are updated to be NaN, even though valid (non NaN) gradients are used. I checked to make sure the grads weren't too big or small, and at all times they were between 0 and 1 (which seems strange, no negatives!). I'm not sure if it's an issue in the Adam optimizer.

Another issue I had was for taking the mean of a tensor in the loss function, the division denominator was > 65_000 so it was inf, therefore the loss was always 0, which I fixed by simply clipping the denominator to 65_000 (bad solution I know), but that's a well understood issue for another time.

**Comment by jafioti** [comment]
Another interesting thing I noticed is when I use full precision I do get negative gradients, though very small (-1.5541445e-8), which doesn't happen on half.

**Comment by ViliamVadocz** [comment]
Ayou using cuda? If so, which compute capability?
You can find this if you run `nvidia-smi --query-gpu=compute_cap --format=csv`

**Comment by jafioti** [comment]
@ViliamVadocz Yeah compute capability is 7.5

**Comment by coreylowman** [comment]
As far as the optimizers - doing all the computations directly in f16 was intentional. It should match pretty closely with what happens if you run optimizers in pytorch in f16.

For AMP they will likely convert the f16 values into f32, do the computations, and then cast back. We could do this for f16 too, but it seemed simpler to match pytorch behavior (as far as I could tell).

**Comment by coreylowman** [comment]
Okay the mean issue should be resolved. Optimizers I guess is expected, but should be better once AMP is added.

Going to close this for now.
